### ISSUE TITLE
Trainer has generally poor performance finetuning Language models

### ISSUE THREAD

**Issue by dhairyadalal** [PyTorchLightning/pytorch-lightning#8393]
I'm not sure if I'm setting up my models in Pytorch-lightning incorrectly but I've found that generally, pytorch-lightning has poor performance compared to a manual training loop and the Huggingface Trainer for my NLP models.

For example, I'd like to finetune RoBERTa for a multi-class classification problem (Clinc 151). Using the transformers implementation out of the box I get the following accuracy scores on the test set:
Pytorch-Lightning Trainer: 0.004
Manual Training Loop: .892
Huggingface Trainer: .907

The manual training loop uses Pytorch's mixed-precision training and the same optimizer/scheduler setup as the one in Pytorch-lightning. Given the stark difference in performance I'm assuming I'm writing my pytorch-lightning model incorrectly but not sure what's wrong with my implementation. 

The code to reproduce the results above can be found in the Colab below. 
CoLab example: https://colab.research.google.com/drive/14r4LCXLFI5DTaRC0izCqwv66FDtDCtSe?usp=sharing

Here's the code specific to Pytorch-lightning. 

```
class LightningIntentModel(pl.LightningModule):

    def __init__(self):
        
        super().__init__()
        self.model = RobertaForSequenceClassification.from_pretrained("roberta-base", 
                                                                      num_labels=NUM_LABELS)
    
    def forward(self, batch, use_labels: bool = True) -> dict:
        if use_labels:
            return self.model.forward(input_ids=batch["input_ids"], 
                                     attention_mask=batch["attention_mask"], 
                                     labels = batch["labels"])    
        else:
            return self.model.forward(input_ids=batch["input_ids"], attention_mask=batch["attention_mask"])
    
    def training_step(self, batch, batch_idx):
        output = self.forward(batch, use_labels=True)

        self.log('train_loss', output["loss"])
        return output["loss"]

    def validation_step(self, batch, batch_idx):
        output = self.forward(batch, use_labels=True)

        _, yhat = torch.max(output["logits"], dim=1)
        val_acc = accuracy_score(yhat.detach().tolist(), batch["labels"].detach().tolist())
        
        self.log('val_loss', output["loss"], prog_bar=True)
        self.log('val_acc', val_acc, prog_bar=True)

        return output["loss"]
            
    def configure_optimizers(self):
        optimizer = AdamW(model.parameters(), lr=1e-3, weight_decay=0.01)
        num_training_steps = 2390
        lr_scheduler = get_scheduler(
            "linear",
            optimizer=optimizer,
            num_warmup_steps=500,
            num_training_steps=num_training_steps
        )

        return [optimizer], [lr_scheduler]


trainer = pl.Trainer(max_epochs=10, 
                     precision=16,
                     gpus=1,
                     default_root_dir="./")
trainer.fit(lightning_model, train_dataloader=train_dl, val_dataloaders=val_dl)
```

**Comment by ananyahjha93** [comment]
@SeanNaren Have you experienced a behavior like this?

**Comment by SeanNaren** [comment]
hey @dhairyadalal thanks for the very detailed reproduce!

I wonder if this is to do with the learning scheduler, since it's supposed to be called every step rather than every epoch (which is the default for Lightning). Could you try modify `configure_optimizers` like such:

```python
def configure_optimizers(self):
    optimizer = AdamW(model.parameters(), lr=5e-5, weight_decay=0.01)
    num_training_steps = 2390
    lr_scheduler = get_scheduler(
        "linear",
        optimizer=optimizer,
        num_warmup_steps=500,
        num_training_steps=num_training_steps
    )

    return [optimizer], [{'scheduler': lr_scheduler, 'interval': 'step'}]
```

and see if this works? It does seem like an optimization issue which is causing the low val accuracies!

On a side note, I'm hoping we can automate that `num_training_step` for you. The function exists in a few places (see lightning-transformers https://github.com/PyTorchLightning/lightning-transformers/blob/master/lightning_transformers/core/model.py#L56. We should try to upstream this to lightning for this use case :)

**Comment by dhairyadalal** [comment]
Unfortunately no improvement with the scheduler set to a step interval. It's worth noting that removing the scheduler entirely from the manual loop doesn't really affect performance either. However, removing mixed-precision training does result in convergence failure. Perhaps there's an issue with the mixed-precision setup in my code for pytorch-lightning implementation? 

https://colab.research.google.com/drive/14r4LCXLFI5DTaRC0izCqwv66FDtDCtSe?usp=sharing#scrollTo=rnLBpgp93djY

**Comment by SeanNaren** [comment]
I made a few adjustments (and moved it to a script) and it seems to give me a reasonable accuracy!

First thing when I moved it to a script I noticed in `configure_optimizers` the reference to `model` isn't `self.model`. Not sure if this was something weird but I changed it so the `configure_optimizers` function looks like this:

```python
    def configure_optimizers(self):
        optimizer = AdamW(self.model.parameters(), lr=5e-5, weight_decay=0.01)
        num_training_steps = 2390
        lr_scheduler = get_scheduler(
            "linear",
            optimizer=optimizer,
            num_warmup_steps=500,
            num_training_steps=num_training_steps
        )

        return [optimizer], [{'scheduler': lr_scheduler, 'interval': 'step'}]
```

I also noted in your test code you set the lightning module to train mode, which I changed to eval

```python
device = "cuda:0"
lightning_model.eval()
lightning_model.to(device)
```

I now see an accuracy of 0.9. Could you confirm these changes assist?

**Comment by dhairyadalal** [comment]
@SeanNaren Thanks that seemed to work!
